### PR TITLE
Added Error Handling to the AsyncTask

### PR DESCRIPTION
**Pull request by roeezolantz**
Added ErrorHandler to the asyncTask by adding failedToFetchRiders func to the interface, then every implementor will be able to handle errors differently.
Aslo added Status-Codes enum.

### CODE REVIEW

**Comment by soryy708** [comment]
About source control:

1. Break your changes in to seperate commits.
When your description says:
"Added blah blah"
"Also added blah blah"
Clearly, this should be at least 2 seperate commits.

2. Give commits a short & meaningful name.
No need for descriptions here. The diff should speak for itself. If it doesn't break it down to multiple commits.

3. Commits should be written in present tense (think: If I revert a commit, does it make sense?
Revert add ErrorHandler to asyncTask
Revert added ErrorHandler to asyncTask
Which makes more sense?

4. Look at your diff before you commit. Are there things unrelated to the commit message there?
What's the point in bundling a whitespace change with a feature implementation?

5. When you're breaking your changes in to seperate commits, keep in mind that each commit must keep the code compilable.
This is important if it's ever reversed.

File: app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java [hunk 1]
```diff
@@ -9,6 +9,7 @@
 import android.net.Uri;
 import android.os.Bundle;
 import android.support.design.widget.FloatingActionButton;
+import android.support.design.widget.Snackbar;
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java:12]: You're not using Snackbar

File: app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java [hunk 2]
```diff
@@ -45,7 +48,7 @@
 import java.util.Calendar;
 
 public class HitchRideActivity extends AppCompatActivity
-    implements ServerConnection.OnRidesRetrievedListener,
+        implements ServerConnection.OnRidesRetrievedListener,
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java:51]: Pointless whitespace change. Shouldn't be in the commit. (Maybe in a seperate commit)

File: app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java [hunk 3]
```diff
@@ -98,7 +101,7 @@ public void filterRides(View view) {
 
         Ride filterRide;
         if (mRideDestination == null) {
-             filterRide = new Ride(null, mRideTime, mRideOriginID, null, "", null);
+            filterRide = new Ride(null, mRideTime, mRideOriginID, null, "", null);
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java:104]: Pointless whitespace change. Shouldn't be in the commit. (Maybe in a seperate commit)

File: app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java [hunk 4]
```diff
@@ -296,15 +299,27 @@ public void onRidesRetrieved(Ride[] rides) {
             progressBar.setVisibility(View.GONE);
         }
 
-        if (rides.length == 0) {
+        if (rides == null || rides.length == 0) {
             TextView noResultsTextView = (TextView)
-                findViewById(R.id.noResultsTextView);
+                    findViewById(R.id.noResultsTextView);
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java:304]: Pointless whitespace change. Shouldn't be in the commit. (Maybe in a seperate commit)

File: app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java [hunk 5]
```diff
@@ -296,15 +299,27 @@ public void onRidesRetrieved(Ride[] rides) {
...
+            mRidesAdapter.addItems(rides);
         }
+    }
 
-        mRidesAdapter.addItems(rides);
+    @Override
+    public void failedToFetchRiders(String message) {
+        new MaterialDialog.Builder(this)
+                .content(message)
+                .title("ארעה שגיאה בזמן ביצוע הסריקה")
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java:318]: Hard coded string. See: [String Resources](https://developer.android.com/guide/topics/resources/string-resource.html)

File: app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java [hunk 6]
```diff
@@ -296,15 +299,27 @@ public void onRidesRetrieved(Ride[] rides) {
...
+                .titleGravity(GravityEnum.START)
+                .buttonsGravity(GravityEnum.START)
+                .contentGravity(GravityEnum.START)
+                .positiveText("אישור").show();
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/HitchRideActivity.java:322]: Hard coded string. See: [String Resources](https://developer.android.com/guide/topics/resources/string-resource.html)

File: app/src/main/java/home/climax708/librecarpool/ServerConnection.java [hunk 7]
```diff
@@ -142,8 +166,31 @@ protected void onPreExecute() {
...
         protected void onPostExecute(Ride[] rides) {
-            if (mListener != null)
-                mListener.onRidesRetrieved(isCancelled() ? new Ride[0] : rides);
+            switch(getFetchStatus()) {
+                case finished:
+                    if (mListener != null)
+                        mListener.onRidesRetrieved(rides);
+                    break;
+                case unknownexception:
+                    mListener.failedToFetchRiders("שגיאה לא צפויה קרתה, אנא נסה שנית מאוחר יותר");
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/ServerConnection.java:175]: Hard coded string. See: [String Resources](https://developer.android.com/guide/topics/resources/string-resource.html)

File: app/src/main/java/home/climax708/librecarpool/ServerConnection.java [hunk 8]
```diff
@@ -142,8 +166,31 @@ protected void onPreExecute() {
...
+                    break;
+                case readingProblem:
+                    mListener.failedToFetchRiders("לא ניתן לאמת את השרת, אנא נסה שנית מאוחר יותר");
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/ServerConnection.java:178]: Hard coded string. See: [String Resources](https://developer.android.com/guide/topics/resources/string-resource.html)

File: app/src/main/java/home/climax708/librecarpool/ServerConnection.java [hunk 9]
```diff
@@ -142,8 +166,31 @@ protected void onPreExecute() {
...
+                    break;
+                case unreachableUrl:
+                    mListener.failedToFetchRiders("לא ניתן לאמת את השרת, אנא נסה שנית מאוחר יותר");
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/ServerConnection.java:181]: Hard coded string. See: [String Resources](https://developer.android.com/guide/topics/resources/string-resource.html)

File: app/src/main/java/home/climax708/librecarpool/ServerConnection.java [hunk 10]
```diff
@@ -142,8 +166,31 @@ protected void onPreExecute() {
...
+                    break;
+                case canceled:
+                    mListener.failedToFetchRiders("הפעולה בוטלה");
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/ServerConnection.java:184]: Hard coded string. See: [String Resources](https://developer.android.com/guide/topics/resources/string-resource.html)

File: app/src/main/java/home/climax708/librecarpool/ServerConnection.java [hunk 11]
```diff
@@ -142,8 +166,31 @@ protected void onPreExecute() {
...
+                    break;
+                case noInternet:
+                    mListener.failedToFetchRiders("לא נמצא חיבור רשת");
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/ServerConnection.java:187]: Hard coded string. See: [String Resources](https://developer.android.com/guide/topics/resources/string-resource.html)

**Review by MrMindyMind — COMMENTED** [review]
I have reviewed your pull request and here is what I think:

1. Unnecessary use of third-party library for Dialogs. We think the Dialog API provided by Google API is sufficient for our needs, in terms of aesthetics and functionality.
2. On top of that, the use of a Dialog for error feed-back is not necessary. The same thing can be accomplished by using a [Toast](https://developer.android.com/guide/topics/ui/notifiers/toasts.html) or simply displaying a `TextView` instead of results.
3. The use of hard-coded Strings limits the multilingual functionality. It would be a lot better to use [String resources](https://developer.android.com/guide/topics/resources/string-resource.html) (see strings.xml, strings-iw.xml and strings-he.xml).
4. Google advises the check for cancellation within the task itself to avoid unnecessary operations when the task is canceled.
5. Inconsistency with field names (line 23 in ServerConnection ).
6. The use of `AsyncStatus` is somewhat redundant. AsyncTask provides a method to check if the task is canceled (https://developer.android.com/reference/android/os/AsyncTask.html#isCancelled()). When task starts and ends, we are provided with callbacks `onPreExecute` and `onPostExecute` (accordingly) which are sufficient for status handling.
7. Please note that `doInBackground` is executed on a different thread. The use of `setStatus(AsyncStatus.finished)` in line 82 in ServerConnection is prone to bugs as it is possible to check the status before the task has finished.

File: app/src/main/java/home/climax708/librecarpool/ServerConnection.java [hunk 12]
```diff
@@ -175,6 +222,7 @@ protected void onPostExecute(Ride[] rides) {
             }
         }
         Ride[] filteredRidesArray = new Ride[filteredRides.size()];
+
```
Comment (soryy708) [inline app/src/main/java/home/climax708/librecarpool/ServerConnection.java:225]: Pointless whitespace change.

**Comment by roeezolantz** [comment]
1. First, I agree that all the whitespaces changes and lines adding/deleting could be in another commit or something, but I think they are ok and doesnt bother to anything.. Its just some standarts fixes that doesnt change in anything the meaning of the code and thats why I didnt commit them alone.

2. About the hard codes strings - fully agree with your comment, I had to put them in the strings file.

3. 3rd Dialogs libraries - I agree that it could be done with just a toat or something, but I think its not a big deal.. In my opinion, 3rd libraries avoidance is not always good when talking about open source, its just saves some time with a nicer ux, but its your decision.

4. About the underscore : In my opinion, it improves readability of your code because it allows you to look at a snippet and quickly identify out all of the member variables from locals, but i get the point of the inconsistency.

5. Last thing, the meaning of the `AsyncStatus` enum is to save the current status of the fetching and the error. 
According to [this issue, ](https://github.com/Libre-Carpool/Android-Client/issues/7) I was asked to handle the errors by displaying the failure, and add that enum.

Please note me if you want me to fix the realted issues and re-pr them.

**Comment by MrMindyMind** [comment]
About 3rd party libraries - this is not about trying to avoid them as if if they were a bad thing, but rather avoid licenses issues. It is okay to use 3rd party libraries as long as their licenses is okay and matches our requirements.
As for Dialogs - The Dialog API provided by Android is good enough for our requirements, and furthermore, it's flexible enough to implement our own design with ease, that is if we ever need to.

Naming variables is not all about preference and convenience, there's a [naming convention](http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-135099.html#367) that should be followed. While we are in control of how we want to name our variables, it is better to follow the convention so the code is familiar to anyone reading it.
Furthermore, you should try to follow the way variables are named for consistency.

As for "AsyncStatus" enumerator - Please note that enum instances are constants and should follow constants naming convention (all capitals).
The main thing that bothered me with your implementation is the way you used it in doInBackground. You set the status to finished in the beginning of the task - this is prone to bugs as it is executed on a separate thread and fetching the status could be called at any moment - resulting incorrect status at the incorrect time.
Furthermore, your implementation did not solve the issue you were referring to. While you do display the error to the user, you don't pass the status enum as required (by the issue).

**Pull request closed without merging**